src/Arcus.Messaging.Pumps.ServiceBus/AzureServiceBusMessagePump.cs

```diff
@@ -0,0 +1,277 @@
+        protected virtual async Task DeadletterMessageAsync(string lockToken,
+            IDictionary<string, object> messageProperties = null)
+        {
+            Guard.NotNullOrEmpty(lockToken, nameof(lockToken));
+
+            await _messageReceiver.DeadLetterAsync(lockToken, messageProperties);
+        }
+
+        /// <summary>
+        ///     Deadletters the current message
+        /// </summary>
+        /// <param name="lockToken">
+        ///     Token used to lock an individual message for processing. See
+        ///     <see cref="AzureServiceBusMessageContext.LockToken" />
+        /// </param>
+        /// <param name="reason">Reason why it's being deadlettered</param>
+        /// <param name="errorDescription">Description related to the error</param>
+        protected virtual async Task DeadletterMessageAsync(string lockToken, string reason, string errorDescription)
+        {
+            Guard.NotNullOrEmpty(lockToken, nameof(lockToken));
+
+            await _messageReceiver.DeadLetterAsync(lockToken, reason, errorDescription);
+        }
+
+        private MessageHandlerOptions DetermineMessageHandlerOptions(AzureServiceBusMessagePumpSettings messagePumpSettings)
+        {
+            var messageHandlerOptions = new MessageHandlerOptions(HandleReceivedExceptionAsync);
+            if (messagePumpSettings.Options != null)
+            {
+                // Assign the configured defaults
+                messageHandlerOptions.AutoComplete = messagePumpSettings.Options.AutoComplete;
+                messageHandlerOptions.MaxConcurrentCalls = messagePumpSettings.Options.MaxConcurrentCalls ?? messageHandlerOptions.MaxConcurrentCalls;
+
+                Logger.LogInformation("Message pump options were configured instead of Azure Service Bus defaults.");
+            }
+            else
+            {
+                Logger.LogWarning("No message pump options were configured, using Azure Service Bus defaults instead.");
+            }
+
+            return messageHandlerOptions;
+        }
+
+        private async Task<MessageReceiver> CreateMessageReceiverAsync(AzureServiceBusMessagePumpSettings messagePumpSettings)
+        {
+            var rawConnectionString = await messagePumpSettings.GetConnectionStringAsync();
+            var serviceBusConnectionStringBuilder = new ServiceBusConnectionStringBuilder(rawConnectionString);
+
+            MessageReceiver messageReceiver;
+            if (string.IsNullOrWhiteSpace(serviceBusConnectionStringBuilder.EntityPath))
+            {
+                // Connection string doesn't include the entity so we're using the message pump settings
+                if (string.IsNullOrWhiteSpace(messagePumpSettings.EntityName))
+                {
+                    throw new ArgumentNullException(nameof(messagePumpSettings.EntityName), "No entity name was specified while the connection string is scoped to the namespace");
```

Comment:
gigantic  nitpick. technically, it isn't the argument that's `null`. So `ArgumentException` is I think more related.

Reply:
Good point, updated!